File: p2/main.cpp
#include <mutex>
#include <condition_variable>
#include <functional>
class FooBar {
private:
    int n;
    std::mutex m;
    std::condition_variable cv;
    bool printingBar;

public:
    FooBar(int n) {
        this->n = n;
        printingBar = false;
    }

    void oneStep(std::function<void()> printFunc, bool isBar) {
        {
            std::unique_lock lk(m);
            cv.wait(lk, [this, isBar](){return isBar == printingBar;});
        } 
        printFunc();
        {
            std::unique_lock lk(m);
            printingBar = !isBar;
        }
        cv.notify_one();
    }

    void foo(std::function<void()> printFoo) {      
        for (int i = 0; i < n; i++) {
            oneStep(printFoo, false);
        }
    }

    void bar(std::function<void()> printBar) {
        for (int i = 0; i < n; i++) {
            oneStep(printBar, true);
        }
    }
};

File: p2/main.java
class FooBar {
    private int n;
    private boolean printingBar;

    public FooBar(int n) {
        this.n = n;
        this.printingBar = false;
    }

    private void oneStep(Runnable printFunc, boolean isBar) throws InterruptedException {
        synchronized(this){
            while(isBar != printingBar){
                wait();
            }
        } 
        printFunc.run();
        synchronized(this){
            printingBar = !isBar;
            notify();
        }
    }

    public void foo(Runnable printFoo) throws InterruptedException { 
        for (int i = 0; i < n; i++) {       
            oneStep(printFoo, false);
        }
    }

    public void bar(Runnable printBar) throws InterruptedException {
        for (int i = 0; i < n; i++) {       
            oneStep(printBar, true);
        }
    }
}